src/LayerPlan.cpp

```diff
@@ -921,26 +923,56 @@ void LayerPlan::addWall(const LineJunctions& wall, int start_idx, const Settings
     for(size_t point_idx = 1; point_idx < max_index; point_idx++)
     {
         const ExtrusionJunction& p1 = wall[(wall.size() + start_idx + point_idx * direction) % wall.size()];
-        const coord_t line_width = (p0.w + p1.w) / 2; //For lines which in itself vary in width, use the average width of the variable line.
 
-        if (!bridge_wall_mask.empty())
+        if(!bridge_wall_mask.empty())
         {
             computeDistanceToBridgeStart((wall.size() + start_idx + point_idx * direction - 1) % wall.size());
         }
 
-        if (first_line)
+        if(first_line)
         {
             addTravel(p0.p, always_retract);
             first_line = false;
         }
-        if (is_small_feature)
+
+        /*
+        If the line has variable width, break it up into pieces with the
+        following constraints:
+        - Each piece must be smaller than the Maximum Resolution setting.
+        - The difference between the trapezoidal shape of the line and the
+          rectangular shape of the line may not exceed the Maximum Extrusion
+          Area Deviation setting, unless required by the first constraint.
+        Since breaking up a line segment into N pieces (each with averaged
+        width) divides the area deviation by N, we can simply check how many
+        pieces we'd want to get low enough deviation, then check if each piece
+        is not too short at the end.
+        */
+        const coord_t delta_line_width = p1.w - p0.w;
+        const Point line_vector = p1.p - p0.p;
+        const coord_t line_length = vSize(line_vector);
+        const coord_t line_area_deviation = std::abs(delta_line_width) * line_length / 8; //How much the line would deviate from the trapezoidal shape if printed at average width.
+        size_t pieces = std::max(size_t(1), round_up_divide(line_area_deviation, max_area_deviation)); //How many pieces we'd need to stay beneath the max area deviation.
```

Comment:
I have a slight preference for something like
```cpp
const size_t max_pieces_area_deviation = round_up_divide(line_area_deviation, max_area_deviation);
const size_t max_pieces_resolution = size_t(line_length / max_resolution);
const size_t pieces = std::max(size_t(1), std::min(max_pieces_area_deviation, max_pieces_resolution));
```
Results in equal or less computations, and is a bit clearer imo.

Reply:
I don't think this will make a noticeable difference in performance. I also don't feel strongly about which solution is easier to read so I've changed it here: ea3f1e00b53dbeb1e7ba05285e7bb0a7725c926d